cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/UsageDataset.java

```diff
@@ -0,0 +1,260 @@
+   * @return datasets used by applicaionId
+   */
+  public Set<Id.DatasetInstance> getDatasets(Id.Application applicationId) {
+    Id.Program programId = ProgramKeyMaker.getProgramId(applicationId);
+    OrderedPair<Id.Program, Id.DatasetInstance> orderedPair = orderedPairs.get(PROGRAM, DATASET);
+    Map<MDSKey, Boolean> datasetKeys = listKV(orderedPair.makeScanKey(programId), Boolean.TYPE);
+    return orderedPair.getSecond(datasetKeys.keySet());
+  }
+
+  /**
+   * Returns datasets used by an adapter.
+   * @param adapterId adapter
+   * @return datasets used by adapterId
+   */
+  public Set<Id.DatasetInstance> getDatasets(Id.Adapter adapterId) {
+    OrderedPair<Id.Adapter, Id.DatasetInstance> orderedPair = orderedPairs.get(ADAPTER, DATASET);
+    Map<MDSKey, Boolean> datasetKeys = listKV(orderedPair.makeScanKey(adapterId), Boolean.TYPE);
+    return orderedPair.getSecond(datasetKeys.keySet());
+  }
+
+  /**
+   * Returns streams used by a program.
+   * @param programId program
+   * @return streams used by programId
+   */
+  public Set<Id.Stream> getStreams(Id.Program programId) {
+    OrderedPair<Id.Program, Id.Stream> orderedPair = orderedPairs.get(PROGRAM, STREAM);
+    Map<MDSKey, Boolean> datasetKeys = listKV(orderedPair.makeScanKey(programId), Boolean.TYPE);
+    return orderedPair.getSecond(datasetKeys.keySet());
+  }
+
+  /**
+   * Returns streams used by an application.
+   * @param applicationId application
+   * @return streams used by applicaionId
+   */
+  public Set<Id.Stream> getStreams(Id.Application applicationId) {
+    Id.Program programId = ProgramKeyMaker.getProgramId(applicationId);
+    OrderedPair<Id.Program, Id.Stream> orderedPair = orderedPairs.get(PROGRAM, STREAM);
+    Map<MDSKey, Boolean> datasetKeys = listKV(orderedPair.makeScanKey(programId), Boolean.TYPE);
+    return orderedPair.getSecond(datasetKeys.keySet());
+  }
+
+  /**
+   * Returns sterams used by an adapter.
+   * @param adapterId adapter
+   * @return streams used by adapterId
+   */
+  public Set<Id.Stream> getStreams(Id.Adapter adapterId) {
+    OrderedPair<Id.Adapter, Id.Stream> orderedPair = orderedPairs.get(ADAPTER, STREAM);
+    Map<MDSKey, Boolean> datasetKeys = listKV(orderedPair.makeScanKey(adapterId), Boolean.TYPE);
+    return orderedPair.getSecond(datasetKeys.keySet());
+  }
+
+  /**
+   * Returns programs using dataset.
+   * @param datasetInstanceId dataset
+   * @return programs using datasetInstanceId
+   */
+  public Set<Id.Program> getPrograms(Id.DatasetInstance datasetInstanceId) {
+    OrderedPair<Id.DatasetInstance, Id.Program> orderedPair = orderedPairs.get(DATASET, PROGRAM);
+    Map<MDSKey, Boolean> programKeys = listKV(orderedPair.makeScanKey(datasetInstanceId), Boolean.TYPE);
+    return orderedPair.getSecond(programKeys.keySet());
+  }
+
+  /**
+   * Returns programs using stream.
+   * @param streamId stream
+   * @return programs using streamId
+   */
+  public Set<Id.Program> getPrograms(Id.Stream streamId) {
```

Comment:
I think the service that uses the dataset can compose higher level mappings. What do you say?